pulser-simulation/pulser_simulation/operators.py

```diff
@@ -0,0 +1,391 @@
+            # Either the qubit id has been defined twice in an operation:
+            for qids in target_qids:
+                if len(set(qids)) < len(qids):
+                    raise ValueError("Duplicate atom ids in argument list.")
+            # Or it was defined in two different operations
+            raise ValueError(
+                "Each qubit can be targeted by only one operation."
+            )
+        if not target_qids_set.issubset(qubit_ids):
+            raise ValueError(
+                "Invalid qubit names: " f"{target_qids_set - set(qubit_ids)}"
+            )
+    # Generate default operators if no operators were given
+    if operators is None:
+        operators = default_operators(eigenbasis)
+
+    # Build operator
+    op_list = [operators["I"] for j in range(len(qubit_ids))]
+    _qid_index = {qid: i for i, qid in enumerate(qubit_ids)}
+
+    for operator, qids in operations:
+        if qids == "global":
+            return sum(
+                build_operator(
+                    eigenbasis,
+                    qubit_ids,
+                    [(operator, [q_id])],
+                    operators,
+                )
+                for q_id in qubit_ids
+            )
+        if isinstance(operator, str):
+            try:
+                operator = operators[operator]
+            except KeyError:
+                raise ValueError(f"{operator} is not a valid operator")
+        for qubit in qids:
+            k = _qid_index[qubit]
+            op_list[k] = operator
+    return qutip.tensor(list(map(qutip.Qobj, op_list)))
+
+
+@dataclass
+class QutipOperator(Operator):
+    """A class to handle Operators as Qutip.QObj objects."""
+
+    eigenbasis: list[States]
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        check_eigenbasis(self.eigenbasis)
+
+    def to_qutip_operator(self) -> qutip.Qobj:
+        """Outputs the qutip.QObj associated with the Operator."""
+        return sum(
+            [
+                operation[0]
+                * build_operator(
+                    self.eigenbasis,
+                    self.qubit_ids,
+                    [
+                        (
+                            build_1qudit_operator(
+                                self.eigenbasis,
+                                qudit_operation[0].operations,
+                                self.operators_dict,
+                            ),
+                            qudit_operation[1],
+                        )
+                        for qudit_operation in operation[1]
+                    ],
+                    self.operators_dict,
+                )
+                for operation in self.operator_string.operations
+            ]
+        )
+
+    @classmethod
+    def from_qutip_Qobj(
+        cls,
+        qutip_obj: qutip.Qobj,
+        qubit_ids: Collection[QubitId],
+        eigenbasis: list[States],
+    ) -> QutipOperator:
```

Comment:
This is surely not the most elegant way to store it, I can try to avoid storing one operator per qubit and try to store one (operator, [all the qubits it addresses]) I would just like to agree on the global structure before moving forward.

Reply:
I don't know qutip quite well enough to understand the difference between a `Qutip.QObj` and a `QutipOperator`. Given that this part of the code is unrelated the the interface we're trying to define, I don't really feel qualified to comment.